examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+        self._box_loss = RetinaNetBoxLoss(delta)
+        self._num_classes = num_classes
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.cast(y_pred, dtype=tf.float32)
+        box_labels = y_true[:, :, :4]
+        box_predictions = y_pred[:, :, :4]
+        cls_labels = tf.one_hot(
+            tf.cast(y_true[:, :, 4], dtype=tf.int32),
+            depth=self._num_classes,
+            dtype=tf.float32,
+        )
+        cls_predictions = y_pred[:, :, 4:]
+        positive_mask = tf.cast(tf.greater(y_true[:, :, 4], -1.0), dtype=tf.float32)
+        ignore_mask = tf.cast(tf.equal(y_true[:, :, 4], -2.0), dtype=tf.float32)
+        cls_loss = self._cls_loss(cls_labels, cls_predictions)
+        box_loss = self._box_loss(box_labels, box_predictions)
+        cls_loss = tf.where(tf.equal(ignore_mask, 1.0), 0.0, cls_loss)
+        box_loss = tf.where(tf.equal(positive_mask, 1.0), box_loss, 0.0)
+        normalizer = tf.reduce_sum(positive_mask, axis=-1)
+        cls_loss = tf.math.divide_no_nan(tf.reduce_sum(cls_loss, axis=-1), normalizer)
+        box_loss = tf.math.divide_no_nan(tf.reduce_sum(box_loss, axis=-1), normalizer)
+        loss = cls_loss + box_loss
+        return loss
+
+
+def add_l2_regularization(weight, alpha=0.0001):
+    """Utility function to add l2 regularization"""
+
+    def _add_l2_regularization():
+        l2_reg = tf.keras.regularizers.l2(alpha)
+        return l2_reg(weight)
+
+    return _add_l2_regularization
+
+
+"""
+## Setting up training parameters
+"""
+
+model_dir = "retinanet/"
+label_encoder = LabelEncoder()
+
+num_classes = 80
+batch_size = 2
+
+learning_rates = [2.5e-06, 0.000625, 0.00125, 0.0025, 0.00025, 2.5e-05]
+learning_rate_boundaries = [125, 250, 500] + [240000, 360000]
+learning_rate_fn = tf.optimizers.schedules.PiecewiseConstantDecay(
+    boundaries=learning_rate_boundaries, values=learning_rates
+)
+
+"""
+## Initializing and compiling model
+"""
+
+loss_fn = RetinaNetLoss(num_classes)
+model = RetinaNet(num_classes)
+
+
+#  Add L2 regularization to all Conv2d weights
+for weight in model.trainable_weights:
+    if re.search(r"^(conv).*(kernel:0)$", weight.name) is not None:
+        model.add_loss(add_l2_regularization(weight))
+
+optimizer = tf.optimizers.SGD(learning_rate=learning_rate_fn, momentum=0.9)
+model.compile(loss=loss_fn, optimizer=optimizer)
+
+"""
+## Setting up callbacks
+"""
+
+callbacks_list = [
+    tf.keras.callbacks.ModelCheckpoint(
+        filepath=os.path.join(model_dir, "weights" + "_epoch_{epoch}"),
+        monitor="loss",
+        save_best_only=False,
+        save_weights_only=True,
+        verbose=1,
+    )
+]
+
+"""
+## Load the COCO2017 dataset using TensorFlow Datasets
+"""
+
+#  set `data_dir=None` to load the complete dataset
+
+(train_dataset, val_dataset), dataset_info = tfds.load(
+    "coco/2017", split=["train", "validation"], with_info=True, data_dir="data"
+)
+
+"""
+## Setting up tf.data pipeline
```

Comment:
There are various non-obvious steps in this setup, e.g. padding. Make sure to explain it

Reply:
why padding is required?

Reply:
Padding is required for 
 - Images because, the shape of each image is different. I pad each image in the batch to the shape of the biggest image in the batch. This effectively means that each batch of images will have a different `height` and `width`.
 - For boxes and class labels because the number of objects presents in each image is different, and to avoid working with ragged tensors, I decided to pad them and make them rectangular.